pkg/compact/compact.go

```diff
@@ -984,3 +985,50 @@ func (c *BucketCompactor) Compact(ctx context.Context) (rerr error) {
 	level.Info(c.logger).Log("msg", "compaction iterations done")
 	return nil
 }
+
+var _ block.MetadataFilter = &GatherNoCompactionMarkFilter{}
+
+// GatherNoCompactionMarkFilter is a block.Fetcher filter that passes all metas. While doing it, it gathers all no-compact-mark.json markers.
+// Not go routine safe.
+// TODO(bwplotka): Add unit test.
+type GatherNoCompactionMarkFilter struct {
+	logger             log.Logger
+	bkt                objstore.InstrumentedBucketReader
+	noCompactMarkedMap map[ulid.ULID]*metadata.NoCompactMark
+}
+
+// NewGatherNoCompactionMarkFilter creates GatherNoCompactionMarkFilter.
+func NewGatherNoCompactionMarkFilter(logger log.Logger, bkt objstore.InstrumentedBucketReader) *GatherNoCompactionMarkFilter {
+	return &GatherNoCompactionMarkFilter{
+		logger: logger,
+		bkt:    bkt,
+	}
+}
+
+// DeletionMarkBlocks returns block ids that were marked for deletion.
```

Comment:
`// NoCompactMarkedBlocks returns block ids that were marked for no compaction.`